scheduler/executor/workflow_executor.go

```diff
@@ -0,0 +1,349 @@
+				return false
+			}
+			// For idempotent behavior ignore the AlreadyExists error which happens if we try to schedule a launchplan
+			// for execution at the same time which is already available in admin.
+			// This is possible since idempotency gurantees are using the schedule time and the identifier
+			if grpcError := status.Code(err); grpcError == codes.AlreadyExists {
+				logger.Debugf(ctx, "duplicate schedule %+v already exists for schedule", s)
+				return false
+			}
+			w.metrics.FailedExecutionCounter.Inc()
+			logger.Error(ctx, "failed to create execution create request %+v due to %v", executionRequest, err)
+			// TODO: Handle the case when admin launch plan state is archived but the schedule is active.
+			// After this bug is fixed in admin https://github.com/flyteorg/flyte/issues/1354
+			return true
+		},
+		func() error {
+			_, execErr := w.adminServiceClient.CreateExecution(context.Background(), executionRequest)
+			return execErr
+		},
+	)
+	return nil
+}
+
+func (w *workflowExecutor) Run() {
+	ctx := context.Background()
+	schedules, err := w.db.SchedulableEntityRepo().GetAll(ctx)
+	if err != nil {
+		panic(fmt.Errorf("unable to run the workflow executor after reading the schedules due to %v", err))
+	}
+	catchUpTill := time.Now()
+	err = w.CatchUpAllSchedules(ctx, schedules, catchUpTill)
+	if err != nil {
+		panic(fmt.Errorf("unable to catch up on the schedules due to %v", err))
+	}
+
+	// Start the go routine to write the snapshot map to the configured snapshot file
+	go w.CheckPointState(ctx)
+
+	defer logger.Infof(ctx, "Exiting Workflow executor")
+	for true {
+		for _, s := range schedules {
+			funcRef := func(ctx context.Context, schedule models.SchedulableEntity, scheduleTime time.Time) {
+				// If the schedule has been deactivated and then the inflight schedules can stop
+				if !*schedule.Active {
+					return
+				}
+				nameOfSchedule := GetScheduleName(schedule)
+				_ = w.rateLimiter.Take()
+				err := w.fire(ctx, scheduleTime, schedule)
+				if err != nil {
+					logger.Errorf(ctx, "unable to fire the schedule %+v at %v time due to %v", s, scheduleTime, err)
+					return
+				} else {
+					w.snapshot.UpdateLastExecutionTime(nameOfSchedule, scheduleTime)
+				}
+			}
+
+			// Register or deregister the schedule from the scheduler
+			if !*s.Active {
+				w.goGfInterface.DeRegister(ctx, s)
+			} else {
+				err := w.goGfInterface.Register(ctx, s, funcRef)
+				if err != nil {
+					w.metrics.ScheduleRegistrationFailure.Inc()
+					logger.Errorf(ctx, "unable to register the schedule %+v due to %v", s, err)
+				}
+			}
+		}
+		time.Sleep(executorSleepTime * time.Second)
+		schedules, err = w.db.SchedulableEntityRepo().GetAll(ctx)
+		if err != nil {
+			w.metrics.ScheduleReadFailure.Inc()
+			logger.Errorf(ctx, "going to sleep additional %v backoff time due to DB error %v", backOffSleepTime, err)
+			time.Sleep(backOffSleepTime * time.Second)
+		}
+	}
+}
+
+func (w *workflowExecutor) Stop() error {
```

Comment:
this is important. Today your exit is not graceful. You can do that by having a bunch of this background threads using a 

for {
  select {
  }
}
loop. And you can use a Done channel in context to stop the Run command.
So instead of having a Stop, just change run to be

Run(ctx: context.Context)

And delete Stop. I can share examples here